internal/sourceanalysis/sourceanalysis.go

```diff
@@ -0,0 +1,84 @@
+package sourceanalysis
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"golang.org/x/exp/slices"
+	"golang.org/x/vuln/exp/govulncheck"
+
+	"github.com/google/osv-scanner/internal/govulncheckshim"
+	"github.com/google/osv-scanner/internal/output"
+	"github.com/google/osv-scanner/pkg/models"
+)
+
+// vulnsFromAllPkgs returns the flattened list of unique vulnerabilities
+func vulnsFromAllPkgs(pkgs []models.PackageVulns) []models.Vulnerability {
+	flatVulns := map[string]models.Vulnerability{}
+	for _, pv := range pkgs {
+		for _, vuln := range pv.Vulnerabilities {
+			flatVulns[vuln.ID] = vuln
+		}
+	}
+
+	vulnList := []models.Vulnerability{}
+	for _, v := range flatVulns {
+		vulnList = append(vulnList, v)
+	}
+
+	return vulnList
+}
+
+// Run runs the language specific analyzers on the code given packages and source info
+func Run(r *output.Reporter, source models.SourceInfo, pkgs []models.PackageVulns) {
+	vulnsSlice := vulnsFromAllPkgs(pkgs)
+
+	// GoVulnCheck
+	if source.Type == "lockfile" && filepath.Base(source.Path) == "go.mod" {
+		res, err := govulncheckshim.RunGoVulnCheck(filepath.Dir(source.Path), vulnsSlice)
+		if err != nil {
+			// TODO: Better method to identify the type of error and give advice specific to the error
+			r.PrintError(
+				fmt.Sprintf("Failed to run code analysis (govulncheck) on '%s' because %s\n"+
+					"(the Go toolchain is required)\n", source.Path, err.Error()))
+
+			return
+		}
+		gvcResByVulnID := map[string]*govulncheck.Vuln{}
+		for _, v := range res.Vulns {
+			gvcResByVulnID[v.OSV.ID] = v
+		}
+
+		for _, pv := range pkgs {
+			// Use index to keep reference to original element in slice
+			for groupIdx := range pv.Groups {
+				for _, vulnID := range pv.Groups[groupIdx].IDs {
+					gvcVuln, ok := gvcResByVulnID[vulnID]
+					if !ok { // If vulnerability not found, analysis has not been performed, so skip
```

Comment:
If no modules have vulnerable pkgs imported, would this still work? I.e. if does govulncheck ever return OSV in cases where Modules is empty? 

If not, we'll have to check if the OSV has source level info.

Reply:
If it's not imported at all, it does not show up (it's treated as if no analysis is done at all, which is incorrect), however, this is also very niche, since if someone runs `go mod tidy` it removes those false positive entries from the `go.mod` file anyway.

Reply:
Hmm, there's an important distinction here though right -- the project can still import from these modules (and hence not be removed by `go mod tidy`), just not any "vulnerable" ones.

Per 

```
	// For example, a module M with two packages M/p1 and M/p2, where only p1
	// is vulnerable, will appear in this list if and only if p1 is imported by
	// the target source code or binary.
```

It seems like this would be incorrect in such a case?

Reply:
Ah yes, in that case you are right, go mod tidy will keep it, and osv-scanner will show a false positive.

In this case we'll have to do some check of whether the vulnerability is a GO-20xx-xxxx based vuln, and set called to false if it never showed up in the list returned by govulncheck?

Reply:
Yeah, but rather than keying just on the ID, a better approach would be checking if the ecosystem_specific contains Go source information.

Reply:
Ok, should be resolved, there is still some edge cases where multiple version ranges that's being affected are saved in different packages (possibly because the source level information is different between different versions).

This would require the semantic package and checking each affected entry is inside the version range or not. Leaving that as future work to implement when offline mode is implemented, as I believe there'll be a lot of overlap in the version range comparison code.